### PR TITLE
usecases: Update multisig

### PR DESCRIPTION
**Pull request by j-mueller**
* Update the multisig contract to use the new contracts API
* Add the ability to customise the signing process to
'Language.Plutus.Contract.Trace'

### CODE REVIEW

File: plutus-contract/src/Language/Plutus/Contract/Trace.hs [hunk 1]
```diff
@@ -109,8 +111,9 @@ type ContractTrace s e m a = StateT (ContractTraceState s (TraceError e) a) m
 
 data WalletState s =
     WalletState
-        { walletEvents   :: Seq (Event s)
-        , walletHandlers :: Handlers s
+        { walletEvents         :: Seq (Event s)
+        , walletHandlers       :: Handlers s
+        , walletSigningProcess :: SigningProcess
```
Comment (michaelpj) [inline plutus-contract/src/Language/Plutus/Contract/Trace.hs]: This feels to me like it's in the wrong place, shouldn't this be in `plutus-emulator` somwhere?

File: plutus-contract/src/Language/Plutus/Contract/Wallet.hs [hunk 2]
```diff
@@ -124,12 +130,28 @@ addOutputs pk vl tx = tx & over Tx.outputs (pko :) where
...
---   known to the wallet.
-addSignatures :: MonadWallet m => [PubKey] -> Tx -> m Tx
-addSignatures pks tx = foldM WAPI.signTxnWithKey tx pks
+handleTx :: MonadWallet m => SigningProcess -> UnbalancedTx -> m Tx
+handleTx p utx =
+    balanceWallet utx >>= addSignatures p (view T.requiredSignatures utx) >>= WAPI.signTxAndSubmit
+
+-- | The signing process gets a finished transaction and a list of public keys,
+--   and signs the transaction with the corresponding private keys.
+newtype SigningProcess = SigningProcess { addSignatures :: forall m. MonadWallet m => [PubKey] -> Tx -> m Tx }
```
Comment (michaelpj) [inline plutus-contract/src/Language/Plutus/Contract/Wallet.hs]: Looks like an effect handler to me :laughing:

Comment (j-mueller) [reply plutus-contract/src/Language/Plutus/Contract/Wallet.hs]: It totally is an effect handler :D

File: plutus-contract/src/Language/Plutus/Contract/Trace.hs (hunk 1 above)
Comment (j-mueller) [reply plutus-contract/src/Language/Plutus/Contract/Trace.hs]: Yeah true I'll move it to `Wallet.Emuator.Types.WalletState`

Comment (j-mueller) [reply plutus-contract/src/Language/Plutus/Contract/Trace.hs]: The idea of a signing process that's separate from the wallet doesn't exist in the emulator though. Everything is part of `WalletAPI` so I don't think we should make `WalletAPI` even more powerful by adding stuff to it. `Wallet.API.signTxn` even has a note saying "In the future this won't be part of WalletAPI to allow the signing to be handled by a different process".

**Pull request merged**